diff -up nfs-utils-1.2.1/utils/exportfs/exportfs.c.orig nfs-utils-1.2.1/utils/exportfs/exportfs.c
--- nfs-utils-1.2.1/utils/exportfs/exportfs.c.orig	2009-11-04 06:13:56.000000000 -0500
+++ nfs-utils-1.2.1/utils/exportfs/exportfs.c	2010-01-12 07:59:03.730815650 -0500
@@ -13,6 +13,7 @@
 #endif
 
 #include <sys/vfs.h>
+#include <sys/stat.h>
 #include <unistd.h>
 #include <stdlib.h>
 #include <string.h>
diff -up nfs-utils-1.2.1/utils/mount/mount.c.orig nfs-utils-1.2.1/utils/mount/mount.c
--- nfs-utils-1.2.1/utils/mount/mount.c.orig	2010-01-12 07:58:57.697003286 -0500
+++ nfs-utils-1.2.1/utils/mount/mount.c	2010-01-12 08:00:45.274357659 -0500
@@ -24,6 +24,7 @@
 
 #include <unistd.h>
 #include <sys/types.h>
+#include <sys/stat.h>
 #include <stdio.h>
 #include <string.h>
 #include <errno.h>
diff -up nfs-utils-1.2.1/utils/mount/network.c.orig nfs-utils-1.2.1/utils/mount/network.c
--- nfs-utils-1.2.1/utils/mount/network.c.orig	2010-01-12 07:58:57.698003139 -0500
+++ nfs-utils-1.2.1/utils/mount/network.c	2010-01-12 07:59:44.041815690 -0500
@@ -37,6 +37,7 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/wait.h>
+#include <sys/stat.h>
 #include <netinet/in.h>
 #include <rpc/rpc.h>
 #include <rpc/pmap_prot.h>
