diff -up nfs-utils-2.6.2/systemd/auth-rpcgss-module.service.orig nfs-utils-2.6.2/systemd/auth-rpcgss-module.service
--- nfs-utils-2.6.2/systemd/auth-rpcgss-module.service.orig	2022-12-05 16:05:29.432966511 -0500
+++ nfs-utils-2.6.2/systemd/auth-rpcgss-module.service	2022-12-05 16:06:20.096192908 -0500
@@ -7,8 +7,8 @@
 [Unit]
 Description=Kernel Module supporting RPCSEC_GSS
 DefaultDependencies=no
-Before=gssproxy.service rpc-svcgssd.service rpc-gssd.service
-Wants=gssproxy.service rpc-svcgssd.service rpc-gssd.service
+Before=gssproxy.service rpc-gssd.service
+Wants=gssproxy.service rpc-gssd.service
 ConditionPathExists=/etc/krb5.keytab
 ConditionVirtualization=!container
 
