--- nfs-utils-1.1.0/utils/statd/statd.h.orig	2007-05-10 23:40:57.000000000 -0400
+++ nfs-utils-1.1.0/utils/statd/statd.h	2007-07-14 05:55:57.000000000 -0400
@@ -16,10 +16,11 @@
 /*
  * Paths and filenames.
  */
+#define STATD_PATH_XTN "statd/"
 #if defined(NFS_STATEDIR)
-# define DEFAULT_DIR_BASE	NFS_STATEDIR "/"
+# define DEFAULT_DIR_BASE	NFS_STATEDIR "/" STATD_PATH_XTN
 #else
-# define DEFAULT_DIR_BASE	"/var/lib/nfs/"
+# define DEFAULT_DIR_BASE	"/var/lib/nfs/" STATD_PATH_XTN
 #endif
 
 #define DEFAULT_SM_DIR		DEFAULT_DIR_BASE "sm"
--- nfs-utils-1.1.0/utils/statd/statd.man.orig	2007-05-10 23:40:57.000000000 -0400
+++ nfs-utils-1.1.0/utils/statd/statd.man	2007-07-14 05:57:20.000000000 -0400
@@ -24,7 +24,7 @@ reboots.
 For each NFS client or server machine to be monitored,
 .B rpc.statd
 creates a file in
-.BR /var/lib/nfs/sm .
+.BR /var/lib/nfs/statd/sm .
 When starting, it normally runs
 .B sm-notify
 to iterate through these files and notify the
@@ -176,11 +176,11 @@ and send notifications to clients. This 
 of an NFS export from another server.
 
 .SH FILES
-.BR /var/lib/nfs/state
+.BR /var/lib/nfs/statd/sm/state
 .br
-.BR /var/lib/nfs/sm/*
+.BR /var/lib/nfs/statd/sm/*
 .br
-.BR /var/lib/nfs/sm.bak/*
+.BR /var/lib/nfs/statd/sm.bak/*
 .SH SEE ALSO
 .BR rpc.nfsd(8),
 .BR portmap(8)
