diff -up nfs-utils-2.3.1/configure.ac.orig nfs-utils-2.3.1/configure.ac
--- nfs-utils-2.3.1/configure.ac.orig	2018-05-22 13:56:01.906548548 -0400
+++ nfs-utils-2.3.1/configure.ac	2018-05-22 13:57:01.197437109 -0400
@@ -157,7 +157,7 @@ AC_ARG_WITH(rpcgen,
 	    RPCGEN_PATH=$rpcgen_path
 	fi
 	AC_SUBST(RPCGEN_PATH)
-	AM_CONDITIONAL(CONFIG_RPCGEN, [test "$RPCGEN_PATH" = ""])
+	AM_CONDITIONAL(CONFIG_RPCGEN, [test "$RPCGEN_PATH" = "NoBuild"])
 AC_ARG_ENABLE(uuid,
 	[AC_HELP_STRING([--disable-uuid], 
 		[Exclude uuid support to avoid buggy libblkid. @<:@default=no@:>@])],
