diff -up nfs-utils-1.2.0/utils/mount/nfsmount.conf.orig nfs-utils-1.2.0/utils/mount/nfsmount.conf
--- nfs-utils-1.2.0/utils/mount/nfsmount.conf.orig	2009-09-30 11:32:14.148326000 -0400
+++ nfs-utils-1.2.0/utils/mount/nfsmount.conf	2009-09-30 11:35:24.714787000 -0400
@@ -24,12 +24,12 @@
 # All reads and writes to the 'nfsserver.foo.com' server 
 # will be done with 32k (32768 bytes) block sizes.
 #
-#[ NFSMount_Global_Options ]
+[ NFSMount_Global_Options ]
 # This statically named section defines global mount 
 # options that can be applied on all NFS mount.
 #
-# Protocol Version [2,3]
-# Nfsvers=3
+# Protocol Version [2,3,4]
+Nfsvers=3
 # Network Transport [Udp,Tcp,Rdma]
 # Proto=Tcp
 #
diff -up nfs-utils-1.2.0/utils/mount/stropts.c.orig nfs-utils-1.2.0/utils/mount/stropts.c
--- nfs-utils-1.2.0/utils/mount/stropts.c.orig	2009-09-30 11:32:14.214327000 -0400
+++ nfs-utils-1.2.0/utils/mount/stropts.c	2009-09-30 11:34:35.841020000 -0400
@@ -601,7 +601,7 @@ static int nfs_try_mount(struct nfsmount
 
 	switch (mi->version) {
 	case 0:
-		if (linux_version_code() > MAKE_VERSION(2, 6, 31)) {
+		if (linux_version_code() >= MAKE_VERSION(2, 6, 31)) {
 			errno = 0;
 			result = nfs_try_mount_v4(mi);
 			if (errno != EPROTONOSUPPORT)
